# Day 2: Cube Conundrum

## Part 1

> <p>You're launched high into the atmosphere! The apex of your trajectory just barely reaches the surface of a large island floating in the sky. You gently land in a fluffy pile of leaves. It's quite cold, but you don't see much snow. An Elf runs over to greet you.</p>
> <p>The Elf explains that you've arrived at <em>Snow Island</em> and apologizes for the lack of snow. He'll be happy to explain the situation, but it's a bit of a walk, so you have some time. They don't get many visitors up here; <span title="No, the Elf's name is not 'WOPR'. It's Joshua.">would you like to play a game</span> in the meantime?</p>
> <p>As you walk, the Elf shows you a small bag and some cubes which are either red, green, or blue. Each time you play this game, he will hide a secret number of cubes of each color in the bag, and your goal is to figure out information about the number of cubes.</p>
> <p>To get information, once a bag has been loaded with cubes, the Elf will reach into the bag, grab a handful of random cubes, show them to you, and then put them back in the bag. He'll do this a few times per game.</p>
> <p>You play several games and record the information from each game (your puzzle input). Each game is listed with its ID number (like the <code>11</code> in <code>Game 11: ...</code>) followed by a semicolon-separated list of subsets of cubes that were revealed from the bag (like <code>3 red, 5 green, 4 blue</code>).</p>
> <p>For example, the record of a few games might look like this:</p>
> <pre><code>Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green
> Game 2: 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue
> Game 3: 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red
> Game 4: 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red
> Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green
> </code></pre>
> <p>In game 1, three sets of cubes are revealed from the bag (and then put back again). The first set is 3 blue cubes and 4 red cubes; the second set is 1 red cube, 2 green cubes, and 6 blue cubes; the third set is only 2 green cubes.</p>
> <p>The Elf would first like to know which games would have been possible if the bag contained <em>only 12 red cubes, 13 green cubes, and 14 blue cubes</em>?


In [ ]:
from collections.abc import Iterable, Iterator
from dataclasses import asdict, dataclass
from functools import reduce
from operator import mul
from re import Pattern
from typing import Self

from advent23 import Checker, disp_name, get_chk, get_inp
from advent23.blake import Stringer, remap

BAG = "Bag: 12 red, 13 green, 14 blue"

chk = get_chk()
inp = get_inp(2)

check = Checker(chk)
checks = {}

#### Part 1

Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green
Game 2: 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue
Game 3: 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red
Game 4: 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red
Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green



#### Part 2

Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green
Game 2: 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue
Game 3: 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red
Game 4: 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red
Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green



In [ ]:
@dataclass
class CubeGroup:
    red: int
    green: int
    blue: int

    def __gt__(self, other: Self) -> bool:
        for m, n in zip(asdict(self).values(), asdict(other).values(), strict=True):
            if m > n:
                return True
        return False


def g(N: str, P: str = r".+", **kwds) -> dict[str, Stringer]:  # noqa: N803
    """Named capturing group for unpacking into a Stringer.

    Args:
        N: Name of the capturing group.
        P: Pattern to match.
        kwds: Other keyword arguments for `Stringer`.
    """
    return {N: Stringer(r=r"(?P<$N>$P)", N=N, P=P, **kwds)}


def findall(pat: Pattern[str], string: str) -> list[str]:
    """Consistent `findall` since `re.findall` output depends on number of groups.

    Args:
        pat: Pattern to match.
        string: String to match against.
    """
    return [match.group() for match in pat.finditer(string)]


NO_ASSIGN = []


def run_checks(stringer: Stringer, assign: str | list[str] = NO_ASSIGN) -> Stringer:
    """Run checks and return the `Stringer` that passed them.

    Args:
        stringer: Stringer to substitute, compile, and check.
        assign: Checks to assign if existing checks pass.
    """
    pat = remap(stringer)
    disp_name("pattern", pat.pattern)
    for name, check in checks.items():
        if name not in chk:
            continue
        check(pat)
    for check in assign if isinstance(assign, list) else [assign]:
        chk[check] = checks[check](pat)
    return stringer

In [ ]:
checks["all_games"] = lambda p: findall(p, inp["a"])
checks["first_game"] = lambda p: m.group() if (m := p.match(inp["a"])) else None
stringer = run_checks(
    Stringer(r=r"^$game$$", **g("game")), assign=["all_games", "first_game"]
)

#### Pattern

^(?P<game>.+)$


#### All games

['Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green',
 'Game 2: 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue',
 'Game 3: 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red',
 'Game 4: 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red',
 'Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green']

#### First game

Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green


In [ ]:
checks["game_1_sets"] = lambda p: m["sets"] if (m := p.match(inp["a"])) else None
stringer = run_checks(
    stringer | dict(game=r"$game_count$sets", game_count=r"Game\s\d+:\s", **g("sets")),
    assign="game_1_sets",
)

#### Pattern

^Game\s\d+:\s(?P<sets>.+)$


#### Game 1 sets

3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green


In [ ]:
checks["game_1_num"] = lambda p: m["game_num"] if (m := p.match(inp["a"])) else None
stringer = run_checks(
    stringer | dict(game_count=r"Game\s$game_num:\s", **g("game_num", r"\d+")),
    assign="game_1_num",
)

#### Pattern

^Game\s(?P<game_num>\d+):\s(?P<sets>.+)$


#### Game 1 num

1


In [ ]:
games = {
    int(match["game_num"]): [set_.split(", ") for set_ in match["sets"].split("; ")]
    for match in remap(stringer).finditer(inp["a"])
}
chk["games"] = games

#### Games

{1: [['3 blue', '4 red'], ['1 red', '2 green', '6 blue'], ['2 green']],
 2: [['1 blue', '2 green'],
  ['3 green', '4 blue', '1 red'],
  ['1 green', '1 blue']],
 3: [['8 green', '6 blue', '20 red'],
  ['5 blue', '4 red', '13 green'],
  ['5 green', '1 red']],
 4: [['1 green', '3 red', '6 blue'],
  ['3 green', '6 red'],
  ['3 green', '15 blue', '14 red']],
 5: [['6 red', '1 blue', '3 green'], ['2 blue', '1 red', '2 green']]}

In [ ]:
get_color = remap(
    r=r"^$num\s$color.*$$", **g("num", r"\d+"), **g("color", r"[r|g|b]")
).fullmatch

COLORS = {"r": "red", "g": "green", "b": "blue"}


def walk_pulls(pulls: Iterable[str]) -> Iterator[CubeGroup]:
    counts = {color: 0 for color in COLORS.values()}
    for match in [get_color(p) for p in pulls]:
        if not match:
            continue
        counts[COLORS[match["color"]]] += int(match["num"])
    yield CubeGroup(**counts)


bag = next(iter(walk_pulls(BAG.removeprefix("Bag: ").split(", "))))
chk["bag"] = bag

#### Bag

CubeGroup(red=12, green=13, blue=14)

In [ ]:
def validate_games(games: dict[int, list[list[str]]]) -> Iterator[int]:
    for game_num, game in games.items():
        valid = True
        for set_ in game:
            for pull in walk_pulls(set_):
                if pull > bag:
                    valid = False
                    break
        if valid:
            yield game_num


possible = list(validate_games(games))
chk["possible"] = possible

#### Possible

[1, 2, 5]

> <p>In the example above, games 1, 2, and 5 would have been <em>possible</em> if the bag had been loaded with that configuration. However, game 3 would have been <em>impossible</em> because at one point the Elf showed you 20 red cubes at once; similarly, game 4 would also have been <em>impossible</em> because the Elf showed you 15 blue cubes at once.


In [ ]:
chk["a"] = sum(possible)

#### Part 1

8

> If you add up the IDs of the games that would have been possible, you get <code><em>8</em></code>.</p>
>
> <p>Determine which games would have been possible if the bag had been loaded with only 12 red cubes, 13 green cubes, and 14 blue cubes. <em>What is the sum of the IDs of those games?</em></p>
> </article>

## Part 2

> <p>The Elf says they've stopped producing snow because they aren't getting any <em>water</em>! He isn't sure why the water stopped; however, he can show you how to get to the water source to check it out for yourself. It's just up ahead!</p>
> <p>As you continue your walk, the Elf poses a second question: in each game you played, what is the <em>fewest number of cubes of each color</em> that could have been in the bag to make the game possible?</p>
> <p>Again consider the example games from earlier.</p>


In [ ]:
@dataclass
class CubeGroup:
    red: int = 0
    green: int = 0
    blue: int = 0

    @property
    def power(self) -> int:
        return reduce(mul, asdict(self).values())


def get_mins(game: list[list[str]]) -> CubeGroup:
    mins = asdict(CubeGroup())
    for set_ in game:
        for match in [get_color(p) for p in set_]:
            if not match:
                continue
            c = COLORS[match["color"]]
            pull = int(match["num"])
            if pull > mins[c]:
                mins[c] = pull
    return CubeGroup(**mins)


def walk_games(games: dict[int, list[list[str]]]) -> Iterator[CubeGroup]:
    for game in games.values():
        yield get_mins(game)


walk = walk_games(games)


def listify(walk: CubeGroup) -> list[int]:
    return list(asdict(walk).values())


chk["game_one_fewest"] = listify(next(walk))

#### Game one fewest

[4, 2, 6]

> In game 1, the game could have been played with as few as 4 red, 2 green, and 6 blue cubes. If any color had even one fewer cube, the game would have been impossible.</li>
>
> <pre><code>Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green</code></pre>


In [ ]:
chk["game_two_fewest"] = listify(next(walk))

#### Game two fewest

[1, 3, 4]

> Game 2 could have been played with a minimum of 1 red, 3 green, and 4 blue cubes.</li>
>
> <pre><code>Game 2: 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue</code></pre>


In [ ]:
chk["game_three_fewest"] = listify(next(walk))

#### Game three fewest

[20, 13, 6]

> Game 3 must have been played with at least 20 red, 13 green, and 6 blue cubes.</li>
>
> <pre><code>Game 3: 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red</code></pre>


In [ ]:
chk["game_four_fewest"] = listify(next(walk))

#### Game four fewest

[14, 3, 15]

> Game 4 required at least 14 red, 3 green, and 15 blue cubes.</li>
>
> <pre><code>Game 4: 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red</code></pre>


In [ ]:
chk["game_five_fewest"] = listify(next(walk))

#### Game five fewest

[6, 3, 2]

> Game 5 needed no fewer than 6 red, 3 green, and 2 blue cubes in the bag.</li>
>
> </ul>
> <pre><code>Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green</code></pre>
> <p>The <em>power</em> of a set of cubes is equal to the numbers of red, green, and blue cubes multiplied together.


In [ ]:
minimum_power = [game.power for game in walk_games(games)]
chk["minimum_power"] = minimum_power
chk["b"] = sum(minimum_power)

#### Minimum power

[48, 12, 1560, 630, 36]

#### Part 2

2286

> The power of the minimum set of cubes in game 1 is <code>48</code>. In games 2-5 it was <code>12</code>, <code>1560</code>, <code>630</code>, and <code>36</code>, respectively. Adding up these five powers produces the sum <code><em>2286</em></code>.</p>
>
> <p>For each game, find the minimum set of cubes that must have been present. <em>What is the sum of the power of these sets?</em></p>
> </article>
